Upgrade To THREE 0.166.0, upgrade to Expo SDK 51

markrickert opened the pull request:
This PR is the culmination of a few weeks of work and experimentation around the `expo-three` and `three.js` ecosystems. @Jpoliachik and I were able to get many more examples running in the example app and lots of upgrades to bring this library back into line with the latest expo standards and SDK.

Here are the updates in no particular order:

1. The example app is set up to use the source code of `expo-three` instead of the build directory, making developing on the example app much easier without the build service running.
2. Updated example app to Expo SDK 51 and added section in readme for how to run the example app.
3. Updated `expo-three` to use the latest `expo-module-scripts`
4. Updated to the latest version of THREE (0.166.0) and fixed incompatibilities and deprecations in the examples. There was a [global lighting system update](https://discourse.threejs.org/t/updates-to-lighting-in-three-js-r155/53733) in r155 so some work had to be done to get the lighting system to get the examples working again. 
5. Added a rudimentary stats panel to the example
6. Fixed the rudimentary file extension matching (it was just looking at the end of the url) with a more robust url matcher that takes parameters in the url into account. With tests.
7. Added orbit controls demo with `expo-three-orbit-controls` (had to patch because it still relies on an old version of THREE).
8. Added lots more demos

The example app works on android and web, but I haven't tested on iOS because the simulator doesn't support web-gl.

| iOS | Android | Web |
| --- | --- | --- |
| 🤷‍♂️ | ✅ | ✅ |

All tests pass (including the new ones!)
Linter passes

Please let me know what else needs to be done to get this merged!

This PR will Close: #164, #199, #212, #238, #270, #287, #297, #298, & #299

Thanks to @Jpoliachik for his help in getting this branch working and for porting over some THREE web examples over from the web examples.

markrickert committed:
chore(example): Upgrade example app to use latest expo SDK

markrickert committed:
Another attempt at the example

markrickert committed:
Fix metro config for png files and url matchers in loadAsync

markrickert committed:
Move the example App.tsx back into place

markrickert committed:
Update THREE to version 0.162.0

markrickert committed:
Clean up more of the example code

markrickert committed:
Add react-native-reanimated

markrickert committed:
Aphpabetize devDependencies

(this was automatically done by yarn)

markrickert committed:
Upgrade yarn deps to latest

markrickert committed:
fix: Properly match the urls in loadAsync. With tests.

markrickert committed:
Got tests working!

Jpoliachik committed:
add expo-router to example app

markrickert committed:
Get the example app working on the web.

Jpoliachik committed:
updates after merge conflicts

Jpoliachik committed:
fix to bundle with metro

Jpoliachik committed:
add robot animation example

markrickert committed:
Add save refresh to robot

markrickert committed:
Ensure timeout refs are used in all examples

markrickert committed:
Add orbit controls to the demo Robot.

markrickert committed:
Adds an example of a stats panel to calculate geometry and FPS

markrickert committed:
Ensure stats are only reported in __DEV__ mode

markrickert committed:
Update the readme for `expo-asset` local file loading.

markrickert committed:
Add expo-gl to Peer Dependencies

Should Fix #164

andrija29 commented:
Great work guys! Let's get this merged please @keith-kurak

joekendal commented:
thanks

adavidson21 commented:
This is great stuff! Would love to see it merged.

sebastian-nowak commented:
Any updates on this? This really needs to be merged, `expo-three` is too far behind the current `three` version

Benmidi commented:
Ooof, I really need this too.

markrickert committed:
Squashed commit of the following:

commit 91187fde984c5a5663cc6b6d8e04f8a4c81299b2
Author: lukmccall <kosmatylukasz@gmail.com>
Date:   Mon Apr 29 13:27:22 2024 +0200

    Bump expo-three in example

commit 7da9e4910c3faa180c4e1e9ea858a4db00f6ffc4
Author: lukmccall <kosmatylukasz@gmail.com>
Date:   Mon Apr 29 13:25:05 2024 +0200

    Publish expo-three@7.0.1

commit f716e0928b3e633386dc8419eb6407fd5c855232
Author: Łukasz Kosmaty <kosmatylukasz@gmail.com>
Date:   Mon Apr 29 13:21:36 2024 +0200

    Upgrade `@expo/browser-polyfill` to 1.0.1 (#302)

markrickert committed:
Ran: npx expo install --fix

markrickert committed:
Upgrade ThreeJS to 0.165.0

markrickert committed:
Bump version in the app.json file

markrickert committed:
Add a rainbow effect to the flower field.

markrickert committed:
Clean up some examples and add a generic template file for making new examples

markrickert committed:
Add two new examples: GCode loader and interactive buffer geometry

markrickert committed:
Remove the spinning cubes example for now.

markrickert commented:
I've resolved all the merge conflicts and upgraded ThreeJS to the latest `0.165.0`.

I also added an example of using the `GCodeLoader` 🚢 

![Screenshot 2024-06-12 at 5 50 56 PM](https://github.com/expo/expo-three/assets/139261/42f02a29-2a27-43c8-b440-3d2cecc5874c)

Jpoliachik committed:
add back spinning cubes example with head model removed for simplicity

Jpoliachik committed:
add head model .obj .mtl in separate example

davidlumley commented:
Any updates on what is required for this to be merged?

joekendal commented:
Lgtm

joekendal commented:
@EvanBacon @tsapeta @lukmccall

joekendal commented:
@markrickert I have tested your PR to address #297 

**package.json**
```json
"expo-three": "github:markrickert/expo-three.git#dev",
```

It does not seem to resolve the error: *Failed to getSize of file:///var/mobile/Containers/Data/Application/...*

markrickert committed:
Add back bouncing ball(s) example

markrickert committed:
Update three to 166, expo-module-scripts and example to 51

markrickert committed:
Get tests and linter working

markrickert committed:
Add an image picker to the bouncing balls example.

Josema commented:
I am having some problems instaling the branch repository.

`Unable to resolve "expo-three" from "src/components/App.js"`

```
"expo": "~51.0.17",
"expo-three": "markrickert/expo-three#dev",
"three": "^0.166.0"
```

I also tried installing it this way:
`"expo-three": "git+https://github.com/markrickert/expo-three.git#dev",` 

But it says:
git@github.com: Permission denied (publickey).

tsapeta commented on example/metro.config.js:
```diff
@@ -1,47 +1,40 @@
-      return acc;
-    }, {}),
-  },
+config.resolver.assetExts.push('png');
+config.resolver.sourceExts = config.resolver.sourceExts.filter(
+  (ext) => ext !== 'png'
+);
```
Isn't png already an asset extension?

markrickert replied:
Yup, this was left in accidentally. Resolved.

tsapeta commented on .eslintrc.js:
```diff
@@ -1,2 +1,5 @@
+  root: true,
+  extends: ['universe/native', 'universe/web'],
+  ignorePatterns: ['build'],
+};
```
Why do we need a custom config?

markrickert replied:
Turns out we didnt! I deleted the file and the next time i ran `yarn` it was recreated from the template. linting still works 🥳

tsapeta commented on example/plugins/withFlipperDisabled.ts:
Expo doesn't support Flipper and it's also officially deprecated. As far as I remember the app template doesn't include it at all, so why do we need a plugin that disables it?

markrickert replied:
Deleted!

tsapeta reviewed (commented):
Hey there!
First of all, I'm sorry that this PR has been left alone 😅 I think this happened mostly due to a large scope of changes, including ones that are probably unnecessary and/or could be split into smaller pieces so it's easier for us to review it.

I did the first pass of review, mostly just questions so I can better understand the motivation behind those changes.

I really appreciate the time and effort you guys took to make this!

markrickert committed:
Addressing issues from review.

markrickert commented:
@tsapeta Thanks for the review. I'm new to working on expo modules and i'm just trying to push this library forward. I've addressed your comments in a new commit and i'll comment or resolve them here on github too.

markrickert commented on package.json:
```diff
@@ -17,10 +17,14 @@
     "clean": "expo-module clean",
     "test": "expo-module test",
-    "lint": "expo-module lint",
+    "lint": "eslint src --ext .js,.ts,.tsx",
```
You're right. I switched it back. I did this while trying to figure out the exact versions of eslint and other dependencies we needed to get the linter passing.

markrickert committed:
Deleted webpack config as metro handles this for us now.

markrickert commented on example/webpack.config.js:
```diff
@@ -0,0 +1,15 @@
+const createConfigAsync = require('@expo/webpack-config');
```
You're right. I deleted this file.

markrickert commented on package.json:
```diff
@@ -46,24 +50,41 @@
+    "expo-file-system": "~16.0.9",
+    "expo-gl": "~13.6.0",
+    "expo-module-scripts": "3.5.1",
+    "expo-modules-core": "1.12.11",
```
Understood. I removed it without consequence. Thanks for the tip!

markrickert commented on example/plugins/withBuildProperties.ts:
```diff
@@ -0,0 +1,20 @@
+
+/**
+ *
+ */
```
Deleted this file and moved the build properties to the main app config file.

markrickert commented on package.json:
```diff
@@ -46,24 +50,41 @@
+    "expo": "^51.0.17",
+    "expo-asset": "~9.0.2",
+    "expo-file-system": "~16.0.9",
+    "expo-gl": "~13.6.0",
```
I was able to get rid of `expo-gl` but all the others need to b there for tests to pass :D

whygolfalan commented:
How is this looking? Could really use it!

tsapeta reviewed (approved):
Looking good now, thank you! ❤️

tsapeta commented:
It's been published in version `8.0.0` 🎉 Could someone confirm if everything is working fine at first glance? I'll then promote this version to `latest` on npm.

Josema commented:
> It's been published in version `8.0.0` 🎉 Could someone confirm if everything is working fine at first glance? I'll then promote this version to `latest` on npm.

It is working for me. 👍🎉

```
"expo": "~51.0.18",
"expo-gl": "~14.0.2",
"expo-three": "^8.0.0",
"three": "^0.166.0"
 ```

markrickert commented:
@tsapeta I just checked in a new app, adding `expo-gl`, `expo-three`, and `three` at the versions specified in the comment above and it loaded in the Gcode example perfectly!

Thanks for getting this over the finish line! 🥳
